src/Microsoft.OData.Core/Batch/ODataBatchWriter.cs

```diff
@@ -231,13 +232,14 @@ public Task WriteStartChangesetAsync()
         /// <param name="changesetId"> The change set Id of the batch request. Cannot be null.</param>
         /// <returns>A task instance that represents the asynchronous write operation.</returns>
         /// <exception cref="ODataException">Thrown if the <paramref name="changesetId"/> is null.</exception>
-        public Task WriteStartChangesetAsync(string changesetId)
+        public async Task WriteStartChangesetAsync(string changesetId)
         {
             ExceptionUtils.CheckArgumentNotNull(changesetId, "changesetId");
 
             this.VerifyCanWriteStartChangeset(false);
-            return TaskUtils.GetTaskForSynchronousOperation(() => this.WriteStartChangesetImplementation(changesetId))
-                .FollowOnSuccessWith(t => this.FinishWriteStartChangeset());
+            await this.WriteStartChangesetImplementationAsync(changesetId)
+                .FollowOnSuccessWith(t => this.FinishWriteStartChangeset())

```

Comment:
Out of curiosity, what's the difference between `FollowOnSuccessWithTask` and `FollowOnSuccessWith`

Reply:
Are there async versions of `this.FinishWriteStartChangeSet()`?

Reply:
`FollowOnSuccessWithTask` supports delegate that return another Task instance while `FollowOnSuccessWith` doesn't. As an illustration consider the following two usages:
```csharp
Task.Factory.StartNew(() =>
{
    // Do work
    return new Random().Next(1, 100);
}).FollowOnSuccessWith((antecedentTask) =>
{
    Console.WriteLine("Result: {0}", antecedentTask.Result);
});

// vs.

Task.Factory.StartNew(() =>
{
    // Do work
    // return new Random().Next(1, 100);
}).FollowOnSuccessWithTask((antecedentTask) =>
{
    return Task.Factory.StartNew(() => Console.WriteLine("Result: {0}", antecedentTask.Result));
});
```

Reply:
There's no async version of `FinishWriteStartChangeSet`. The private method currently doesn't do any writing. It's an implementation detail so it'd be easy to introduce an async equivalent if the synchronous version evolved to doing any writing.